tests/tensorflow2/test_error_handling.py

```diff
@@ -0,0 +1,211 @@
+    return f"{out_dir}/tmp.log"
+
+
+@pytest.fixture
+def keras_callback_error_message():
+    return "If this Keras callback error causes the test to fail, the error handler failed to catch the error!"
+
+
+@pytest.fixture
+def layer_callback_error_message():
+    return "If this layer callback error causes the test to fail, the error handler failed to catch the error!"
+
+
+@pytest.fixture
+def custom_configuration_error_message():
+    return "This error should have been thrown and not caught by the error handler!"
+
+
+@pytest.fixture
+def hook_class_with_keras_callback_error(keras_callback_error_message):
+    class HookWithBadKerasCallback(Hook):
+        def __init__(self, keras_error_message=keras_callback_error_message, *args, **kwargs):
+            super().__init__(*args, **kwargs)
+            self.keras_callback_error_message = keras_error_message
+
+        @error_handler.catch_smdebug_errors()
+        def on_train_batch_begin(self, batch, logs=None):
+            raise RuntimeError(self.keras_callback_error_message)
+
+    return HookWithBadKerasCallback
+
+
+@pytest.fixture
+def hook_class_with_layer_callback_error(layer_callback_error_message):
+    class HookWithBadLayerCallback(Hook):
+        def __init__(self, layer_error_message=layer_callback_error_message, *args, **kwargs):
+            super().__init__(*args, **kwargs)
+            self.layer_callback_error_message = layer_error_message
+
+        def _wrap_model_with_input_output_saver(self):
+            def _get_layer_call_fn_error(layer):
+                old_call_fn = layer.call
+
+                @error_handler.catch_smdebug_errors(
+                    return_type="layer_call", old_call_fn=old_call_fn
+                )
+                def call(inputs, *args, **kwargs):
+                    raise RuntimeError(self.layer_callback_error_message)
+
+                return call
+
+            for layer in self.model.layers:
+                layer.call = _get_layer_call_fn_error(layer)
+
+    return HookWithBadLayerCallback
+
+
+@pytest.fixture
+def hook_class_with_keras_and_layer_callback_error(
+    out_dir,
+    hook_class_with_keras_callback_error,
+    hook_class_with_layer_callback_error,
+    keras_callback_error_message,
+    layer_callback_error_message,
+):
+    class HookWithBadKerasAndLayerCallback(
+        hook_class_with_keras_callback_error, hook_class_with_layer_callback_error
+    ):
+        def __init__(self, *args, **kwargs):
+            super().__init__(*args, **kwargs)
+
+    return HookWithBadKerasAndLayerCallback
+
+
+@pytest.fixture
+def profiler_config_path(config_folder):
+    return os.path.join(config_folder, "test_tf2_profiler_config_parser_by_time.json")
+
+
+@pytest.fixture(autouse=True)
+def set_up(out_dir, stack_trace_filepath):
+    logger = get_logger()
+    os.makedirs(out_dir)
+    file_handler = logging.FileHandler(filename=f"{out_dir}/tmp.log")
+    logger.addHandler(file_handler)
+    yield
```

Comment:
Can you explain the purpose of `yield` here?

Reply:
The purpose of this fixture is to simulate setup and teardown functionality for each test in the file. hence the `autouse=True` in `pytest.fixture` - we don't need to manually reference this fixture in each test. 

There are three parts to this design:
- The block of code before `yield` is the setup and gets run before the test is executed.
- `yield` returns execution back to the test. The test then gets executed.
- The block of code after `yield` is the teardown and gets run after the test is executed (regardless of pass or fail).

This is a really interesting [read](https://docs.pytest.org/en/stable/fixture.html#teardown-cleanup-aka-fixture-finalization) that I'd recommend for more detail.